New config

hnformentin opened the pull request:
Solves #191 
Peer-programming with @oyvindeide :1st_place_medal: 

The new configuration proposed in #ert-research.

```
input_data:
  alpha: 2  # remove from config?
  std_cutoff: 0.001  # remove from config?
  observations:
    - W*
    - FOPT
workflow:
    auto_scale:  # optional, mutually exclusive with spearman_correlation
      clustering:  # optional
        hierarchical:  # optional, mutually exclusive with kmeans
            method: complete  # optional
            metric: jensenshannon  # optional
        kmeans:  # optional, mutually exclusive with hierarchical
      pca:  # optional
        threshold: 0.99    
    spearman_correlation:  # optional, mutually exclusive with auto_scale
        clustering:  # optional
           hierarchical:  # optional, mutually exclusive with kmeans
             method: centroid  # optional
             metric: dice  # optional
             criterion: maxclust  # optional
             t: 10  # optional
           kmeans:  # mutually exclusive with hierarchical
             some_option: 2
        pca:  # optional
           threshold: 0.98
```

oyvindeide commented on semeio/workflows/misfit_preprocessor/config.py:
```diff
@@ -251,51 +252,40 @@ def _inject_default_t(config):
 
-_SCALING_SCHEMA = {
+@cs.transformation_msg("Detecting method.")
+def insert_method(layer):
```
Should probably prefix this with `_` and/or add a unit test.

oyvindeide replied:
It might also be good with a docstring in this function to make it clean what the intention is.

hnformentin replied:
In the first comment, I have to think about a unit test.

hnformentin replied:
We expanded `test_config_workflow_invalid` in `test_config.py` for this.

oyvindeide commented on semeio/workflows/misfit_preprocessor/config.py:
```diff
@@ -251,51 +252,40 @@ def _inject_default_t(config):
+            return layer
+        if len(layer) == 2 and layer["method"] in layer:
+            return layer
+        raise ValueError(f"Found method in {layer}. Did not match {layer.keys}.")
```
Perhaps improve this error message? Is a bit cryptic.

hnformentin replied:
We just made it also!

oyvindeide commented on semeio/workflows/misfit_preprocessor/config.py:
```diff
@@ -251,51 +252,40 @@ def _inject_default_t(config):
 
 
-_SCALING_SCHEMA = {
+@cs.transformation_msg("Detecting method.")
```
Maybe improve this message a bit?

hnformentin replied:
Would `Checking workflow schema in the configuration file` be better?

hnformentin replied:
We just made it!

oyvindeide commented on semeio/workflows/misfit_preprocessor/job.py:
```diff
@@ -6,34 +6,47 @@
 from semeio.workflows.spearman_correlation_job.job import spearman_job
 
 
+def _fetch_scaling_parameters(threshold):
```
At this point I think it makes more sense to just update the dict directly with this instead of having this function.

hnformentin replied:
Impressive that you catch these important details even after many times seeing the code!

oyvindeide commented on tests/jobs/misfit_preprocessor/unit/test_config.py:
```diff
@@ -372,3 +389,54 @@ def test_default_observations(observation_keys):
     assert sorted(observation_keys) == sorted(config.snapshot.observations)
+
+
+def test_new_config():
```
Perhaps rename this, or check if the functionality is now covered by the refactored tests? It made a lot of sense when we started this work, but might be surpassed now?

hnformentin replied:
You are right, you even mentioned it in the chat. My bad when rebasing I probably missed it somehow.

oyvindeide commented on tests/jobs/misfit_preprocessor/unit/test_config.py:
```diff
@@ -372,3 +389,54 @@ def test_default_observations(observation_keys):
+@pytest.mark.parametrize("workflow", [{"spearman_correlation": {}, "auto_scale": {}}])
+def test_config_workflow_invalid(workflow):
+    config_data = {"workflow": workflow}
+    #    with pytest.raises(ValueError):
```
Remove commented out line.

oyvindeide commented on tests/jobs/misfit_preprocessor/unit/test_config.py:
```diff
@@ -372,3 +389,54 @@ def test_default_observations(observation_keys):
+    #    with pytest.raises(ValueError):
+    config = misfit_preprocessor.config.MisfitPreprocessorConfig(config_data, ["WWPR"])
+
+    assert not config.valid
```
Perhaps also check that the error message is what we would expect?

hnformentin replied:
The error message in config.errors is empty for this valid config.
Maybe we could try to check the error message in `test_config_workflow_invalid`?

hnformentin replied:
We made a test for that!

oyvindeide commented on semeio/workflows/misfit_preprocessor/config.py:
```diff
@@ -136,106 +136,107 @@ def _inject_default_t(config):
+            cs.MetaKeys.Content: {
+                _SPEARMAN_THRESHOLD: {
+                    cs.MetaKeys.Type: cs.types.Number,
+                    cs.MetaKeys.Description: (
```
Feel free to refactor to: f"String {variable}" syntax if you like (goes for all descriptions)

hnformentin replied:
I found inly this one that needed to be changed :tada:

hnformentin replied:
Would it be ok to change `raise ValidationError` format also?

oyvindeide reviewed (commented):
Good piece of work so far!

hnformentin commented on semeio/workflows/misfit_preprocessor/config.py:
```diff
@@ -136,106 +136,107 @@ def _inject_default_t(config):
-            ).format(maxclust=_MAXCLUST, criterion=_CRITERION),
-            cs.MetaKeys.ElementValidators: (
-                _bounds_validator(lower=0, lower_inclusive=False),
+                "Threshold used when computing primary components of the "
```
Threshold used when computing primary components of the **clusters.**
It was removed some commits ago. Let me know if you think we should add something else :-)

oyvindeide commented on semeio/workflows/misfit_preprocessor/config.py:
```diff
@@ -251,51 +252,40 @@ def _inject_default_t(config):
-                "Threshold used when computing primary components of the "
-            ),
-            cs.MetaKeys.Default: 0.95,
+        "auto_scale": _CLUSTERING_SCHEMA,
```
This configuration should be a bit different, as not all the options will be available when doing `auto_scale`, could perhaps fix this with a context validation.

hnformentin replied:
I tried to understand how context validation would work here, but did not get a clear idea. Maybe we could check  this together.

hnformentin replied:
We proposed a solution for review and discussion!

hnformentin commented:
I addressed some comments (the ones with reactions) and had to force push to fix problems identified by tox in the two commits.
Thanks for the comments so far!

hnformentin commented:
Forced-push to rebase on master.

oyvindeide commented on tests/jobs/misfit_preprocessor/test_integration.py:
```diff
@@ -79,9 +78,13 @@ def test_misfit_preprocessor_passing_scaling_parameters(monkeypatch, test_data_r
+        "workflow": {
+            "spearman_correlation": {"pca": {"threshold": 0.5}},
+        },
+        # "clustering": {"method": "spearman_correlation"},
```
This should probably be deleted.

oyvindeide commented on semeio/workflows/misfit_preprocessor/cluster_config.py:
```diff
@@ -0,0 +1,207 @@
+                    cs.MetaKeys.Description: (
+                        "Scalar threshold for the clustering. When a "
+                        "'{maxclust}' {criterion} is used, the scalar gives "
+                        "the maximum number of clusters to be formed."
```
There should be a space after this sentence (and the three other places this occurs)

hnformentin commented:
This PR came at the time when k_means was being considered for the clustering workflow and the configuration file needed to accommodate more options than the previous version. A proposition of configuration format was agreed with the research and this PR implements it.
We made compromises at several points of the new config file. Two major discussion points were:
1. How to build code blocks when a workflow presents more configuration options than another (in this case hierarchical and auto_scale). The ideal option would be having blocks of codes that build-up on the desired setting of configurations. For example, hierarchical would have all the blocks while auto_scale would only have a couple of configurable blocks. This was not an option with current code base of config_suite. A second alternative was to have separate blocks for each workflow option. We discarded it because it would lead to repetition of code and tests. Our final option was to use the same code block combined with a transformation layer to validate that the options presented in the configuration file are appropriate for a given workflow.
2. How to design the layer transformation: the initial though was to use a tuple of transformations in `cs.MetaKeys.LayerTransformation`. As we understood this was not an option in the current implementation of config_suite, we opted to have a function with `_double_transformation`: one checking than only one method was configures, another checking that all options configured were supported for the chosen workflow.

We agreed that the configuration proposed here will be subject of iterations with users before launching it to production. We understand that the users may have a different perspective and this PR is also meant to foster a discussion.

oyvindeide reviewed (approved):
Good job!
